vmm/src/device_manager.rs

```diff
@@ -895,6 +977,40 @@ impl DeviceManager {
         Ok(device_manager)
     }
 
+    pub fn get_serial_pty(&self) -> Option<File> {
```

Comment:
Do you think you can return `Option<&File>`?
I'm asking because it'd be much nicer if we could avoid `try_clone()`, which effectively means that we could run into some unexpected behaviors if multiple threads were given this `File`, as they could seek/read in parallel, thinking they own the File.

Reply:
I don't really see a good way to avoid a `try_clone` at some point because the epoll usage in vmm/src/lib.rs on lines 364 and 369 require having a `AsRawFd` trait (well without implementing `AsRawFd` for `&File` at least). This also makes me feel that "ownership" for the pty is already somewhat dubious given we give out the raw fd directly anyway. 

Going back to the problem you raise around parallel writes/reads, that is certainly something I'd like to minimize or avoid entirely. I see that the `handle_*_pty` functions would be perfectly happy with reference to file so adding another function for getting the `&File` for those functions (perhaps an aptly named `get_*_pty_as_ref()`?) seems like a potential option.This would mean the handlers would be behind a lock

Reply:
Also see [this](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2168/commits/3dc14ed36be931caaecea220ffccabe86ee58b22#diff-9c91c6326e9d86ef107a59aed3f13a9bb91790f9f4be1668d22cda12a28bc28cR1604) for what I see as a potential problem going forward with this implementation.

Reply:
Ok I see the problem. I don't really like this new version, it's even more complicated without getting rid of the `try_clone()` eventually.
Do you think if `vmm/src/lib.rs` on lines 364 and 369 was expecting an actual `RawFd` instead of `AsRawFd`, it would help simplify the code? I'm suggesting this because I don't really care if we change it.
If it doesn't help, then forget about it, and let's move on with your initial implementation :)

Reply:
Forgetting about this for now =).